test/diag2dwd.jl

```diff
@@ -217,6 +218,317 @@ end
+      function unique_by(acset, column_names::Vector{Symbol})
+
+  Given column names from the same table, return a copy of the acset with
+  duplicate rows removed. Removal of rows is performed with prejudice.
+
+  WARNING: This function does not check if other tables index into the one
+  given. Removal of rows is performed with prejudice.
+
+  See also: [`unique_by!`](@ref).
+
+  # Examples
+  ```julia-repl
+  julia> g = @acset Graph begin
+    V = 2
+    E = 3
+    src = [1,1,2]
+    tgt = [1,1,2]
+  end
+  julia> unique_by(g, :E, [:src, :tgt])
+  ``` 
+  """
+  function unique_by(acset, table::Symbol, columns::Vector{Symbol})
+    acset_copy = copy(acset)
+    unique_by!(acset_copy, table, columns)
+  end 
+
+  """
+      function compose_decapodes(decapodes_vars::Vector{Tuple{NamedDecapode{
+        Any, Any, Symbol}, Vector{Symbol}}}, relation::RelationDiagram)
+
+  Compose a list of decapodes as specified by the given relation diagram.
+
+  The decapodes must be given in the same order as they were specified in the
+  relation.
+
+  State variables (such as the (C,V) given in the head of the following
+  @relation) do not affect the result of a composition.
+
+  # Examples
+  ```julia-repl
+  julia> compose_diff_adv = @relation (C,V) begin
+    diffusion(C, ϕ₁)
+    advection(C, ϕ₂, V)
+    superposition(ϕ₁, ϕ₂, ϕ, C)
+  end
+  julia> compose_decapodes([(Diffusion, [:C, :ϕ]), (Advection, [:C, :ϕ, :V]),
+    (Superposition, [:ϕ₁, :ϕ₂, :ϕ, :C])], compose_diff_adv)
+  ````
+  """
+  function compose_decapodes(decapodes_vars::Vector{Tuple{NamedDecapode{
+    Any, Any, Symbol}, Vector{Symbol}}}, relation::RelationDiagram)
+    # TODO: Calling this function "compose" is something of a misnomer.  e.g.
+    # You could very well give a vector containing a single decapode. Then the
+    # effect of applying the relation would be something like renaming
+    # variables. So this function should really be called oapply_decapodes (or
+    # just oapply if you are willing to import oapply from
+    # Catlab.WiringDiagrams).
+    r = relation
+    copies = decapodes_vars .|> first .|> copy
+    # TODO: We should also check that the number of variables given in the
+    # relation is the same as the number of symbols in the corresponding vector
+    # of Vars.
+
+    # Step -4: Determine the mapping of global ports to local ports.
+    # In a RelationDiagram, this is baked into the order of rows in the Port
+    # table.
+    # (i.e. for each (box,junction) pair, determine whether this corresponds to
+    # the 1st Var of that box, or the 2nd Var of that box, etc.)
+    # tuples (box, indices of rows with that box):
+    box_rows = [(b, incident(r, b, :box)) for b in unique(r[:box])]
+    local_ports = [map(y -> y-idxs[begin]+1, idxs) for (b, idxs) in box_rows]
+    # This is a column that one could hcat to the Ports table.
+    local_ports = local_ports |> flatten |> collect
+
+    # Step -3: Check that the types of all Vars connected by the same
```

Comment:
whenever you need a comment like this, because you have coherent steps to a function, that is a code smell that you should be extracting this block of code into a function. If it is coherent enough to explain with a comment, then it is coherent enough to make a function, docstring, and unit tests.

For example, this is a type checking step that would be a meaningfully useful function outside the scope of composing.